File: Leetcode/python/sol347.py
# https://leetcode.com/problems/top-k-frequent-elements/description/
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count =  Counter(nums)
        freq :list[list[int]] = [[] for i in range(len(nums ) +1)]
        for num, times in count.items():
            freq[times].append(num)

        rslt :list[int] = []
        for i in freq[::-1]:
            if i:
                rslt.extend(i)

        return rslt[:k]


import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = list(Counter(nums).items())
        result = []
        for val, freq in count:
            heapq.heappush(result, (-freq, val))
        rslt = []
        for _ in range(k, 0, -1):
            rslt.append(heapq.heappop(result)[1])
        return rslt
        # if k > len(nums):
        #     return []
        # count =  list(Counter(nums).items())
        # count.sort(key=lambda x:x[1], reverse=True)
        # return [count[i][0] for i in range(k)]

File: Leetcode/cpp/sol347.cpp

// https://leetcode.com/problems/top-k-frequent-elements/description/

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int,int> counter;
        for(int i:nums){
            counter[i]++;
        }
        vector<vector<int>> freq(nums.size()+1);
        for(auto i: counter){
            freq[i.second].push_back(i.first);
        }

        vector <int> rslt;
        int count = 0;

        for(int i = freq.size() - 1; i >= 0; i--){
            if(count == k){
                break;
            }
            for(auto j:freq[i]){
                rslt.push_back(j);
                count+=1;
            }
        }

        return rslt;
    }
};